### PR TITLE
U303/t329 divide physic system

### PR DESCRIPTION
**Pull request by Smaugmuel**
PhysicSystem has been replaced by:
MovementSystem, SpeedLimitSystem, CollisionSystem, MovementPostCollisionSystem.

PhysicsComponent  has been replaced by:
MovementComponent, CollisionComponent, SpeedLimitComponent

### CODE REVIEW

File: Sail/src/Sail/entities/components/CollisionComponent.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include "Component.h"
+#include "..//..//Physics/Octree.h"
+
+class CollisionComponent : public Component<CollisionComponent> {
+public:
+	CollisionComponent() {}
+	~CollisionComponent() {}
+
+	float airDrag = 1.0f;
```
Comment (Praccen) [inline Sail/src/Sail/entities/components/CollisionComponent.h]: Air drag only possible for entities that can collide? Maybe move this to MovementPostCollisionSystem?

File: Sail/src/Sail/entities/components/CollisionPhysicsComponent.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	~CollisionComponent() {}
+
+	float airDrag = 1.0f;
+	float drag = 25.0f;
+	float bounciness = 0.0f;
+	float padding = -1.0f;
+	bool onGround = false;
+
+	std::vector<Octree::CollisionInfo> collisions; //Contains the info for current collisions
+};
```
Comment (Praccen) [inline Sail/src/Sail/entities/components/CollisionPhysicsComponent.h]: Identical to CollisionComponent. Even same class name

File: Sail/src/Sail/entities/systems/Gameplay/ProjectileSystem.cpp [hunk 3]
```diff
@@ -19,8 +17,8 @@ ProjectileSystem::~ProjectileSystem() {
 
 void ProjectileSystem::update(float dt) {
 	for (auto& e : entities) {
-		PhysicsComponent* physComp = e->getComponent<PhysicsComponent>();
-		auto projectileCollisions = physComp->collisions;
+		CollisionComponent* collision = e->getComponent<CollisionComponent>();
+		auto projectileCollisions = collision->collisions;
 		for (auto& collision : projectileCollisions) {
```
Comment (Praccen) [inline Sail/src/Sail/entities/systems/Gameplay/ProjectileSystem.cpp]: "collision" is used as variable name multiple times.

File: Sail/src/Sail/entities/systems/physics/CollisionSystem.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	const bool rayCastCheck(Entity* e, const BoundingBox& boundingBox, float& dt) const;
+	void rayCastUpdate(Entity* e, BoundingBox& boundingBox, float& dt);
+	const bool collisionUpdate(Entity* e, const float& dt);
+	const bool handleCollisions(Entity* e, const std::vector<Octree::CollisionInfo>& collisions, const float& dt);
+	void surfaceFromCollision(Entity* e);
+	float findIntersectionDepth(Entity* e, const Octree::CollisionInfo& collision);
+
+private:
+	Octree* m_octree = nullptr;
+	GameDataTracker* m_gameDataTracker = nullptr;
```
Comment (Praccen) [inline Sail/src/Sail/entities/systems/physics/CollisionSystem.h]: Unused

File: Sail/src/Sail/entities/systems/physics/CollisionSystem.cpp [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+void CollisionSystem::provideOctree(Octree* octree) {
+	m_octree = octree;
+}
+
+void CollisionSystem::update(float dt) {
+	for (auto& e: entities) {
+		auto movement = e->getComponent<MovementComponent>();
+		auto transform = e->getComponent<TransformComponent>();
```
Comment (Praccen) [inline Sail/src/Sail/entities/systems/physics/CollisionSystem.cpp]: Unused

File: Sail/src/Sail/entities/systems/input/GameInputSystem.cpp [hunk 6]
```diff
@@ -80,9 +86,11 @@ void GameInputSystem::processKeyboardInput(const float& dt) {
 
 			// Else do normal movement
 		} else {
-			PhysicsComponent* physicsComp = e->getComponent<PhysicsComponent>();
-			BoundingBoxComponent* playerBB = e->getComponent<BoundingBoxComponent>();
-			AudioComponent* audioComp = e->getComponent<AudioComponent>();
+			auto collision = e->getComponent<CollisionComponent>();
+			auto movement = e->getComponent<MovementComponent>();
+			auto speedLimit = e->getComponent<SpeedLimitComponent>();
+			auto playerBB = e->getComponent<BoundingBoxComponent>();
```
Comment (Praccen) [inline Sail/src/Sail/entities/systems/input/GameInputSystem.cpp]: Unused

File: Sail/src/Sail/entities/components/CollisionComponent.h (hunk 1 above)
Comment (Smaugmuel) [reply Sail/src/Sail/entities/components/CollisionComponent.h]: Moved airDrag to from CollisionComponent to MovementComponent and moved code from CollisionSystem to MovementSystem.

File: Sail/src/Sail/entities/systems/Gameplay/ProjectileSystem.cpp (hunk 3 above)
Comment (Smaugmuel) [reply Sail/src/Sail/entities/systems/Gameplay/ProjectileSystem.cpp]: Renamed first variable to collisionComp

File: Sail/src/Sail/entities/systems/physics/CollisionSystem.h (hunk 4 above)
Comment (Smaugmuel) [reply Sail/src/Sail/entities/systems/physics/CollisionSystem.h]: GameDataTracker member removed from CollisionSystem

File: Sail/src/Sail/entities/systems/physics/CollisionSystem.cpp (hunk 5 above)
Comment (Smaugmuel) [reply Sail/src/Sail/entities/systems/physics/CollisionSystem.cpp]: Removed transform.

File: Sail/src/Sail/entities/systems/input/GameInputSystem.cpp (hunk 6 above)
Comment (Smaugmuel) [reply Sail/src/Sail/entities/systems/input/GameInputSystem.cpp]: Removed playerBB

File: Sail/src/Sail/entities/systems/physics/MovementPostCollisionSystem.cpp [hunk 7]
```diff
@@ -35,5 +35,17 @@ void MovementPostCollisionSystem::update(float dt) {
...
+		// Apply air drag
+		float saveY = movement->velocity.y;
+		movement->velocity.y = 0;
+		float vel = glm::length(movement->velocity);
+
+		if (vel > 0.0f) {
+			vel = glm::max(vel - movement->airDrag * dt, 0.0f);
+			movement->velocity = glm::normalize(movement->velocity) * vel;
+		}
+		movement->velocity.y = saveY;
```
Comment (Praccen) [inline Sail/src/Sail/entities/systems/physics/MovementPostCollisionSystem.cpp]: Do before translation?

**Pull request merged**